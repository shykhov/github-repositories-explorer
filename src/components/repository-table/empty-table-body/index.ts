export * from './empty-table-body';
