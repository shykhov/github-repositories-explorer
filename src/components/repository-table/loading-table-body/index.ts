export * from './loading-table-body';
