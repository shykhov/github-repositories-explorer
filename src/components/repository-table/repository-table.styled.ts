import styled from 'styled-components';
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';

export const StyledTableContainer = styled(TableContainer)`
  max-height: 440px;
`;

export const StyledPaper = styled(Paper)`
  max-width: 800px;
`;

export const StyledTableCell = styled(TableCell)`
  font-weight: bold !important;
`;

export const StyledTable = styled(Table)`
  table-layout: fixed;
`;
