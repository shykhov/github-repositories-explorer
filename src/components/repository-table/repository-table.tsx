import React, { FC, useMemo } from 'react';
import Link from '@material-ui/core/Link';
import isEmpty from 'lodash/isEmpty';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import isNil from 'lodash/isNil';

import { StyledTable, StyledTableContainer, StyledPaper, StyledTableCell } from './repository-table.styled';
import { EmptyTableBody } from './empty-table-body';
import { LoadingTableBody } from './loading-table-body';
import { RepositoryResultData, RepositoryResult } from '../../utils';
import { ContentRenderer } from '../content-renderer';
import { REPOSITORIES_PER_PAGE, REPOSITORY_TABLE_COLUMNS } from '../../constants';

export interface Props {
  error: boolean;
  called: boolean;
  loading: boolean;
  repositoriesData: RepositoryResultData;
  handleChangePage(event: unknown, newPage: number): void;
  page: number;
  repositoryNameSearchParameter: string | null;
  userLoginParameter: string | null;
}

export const RepositoryTable: FC<Props> = props => {
  const {
    handleChangePage,
    page,
    userLoginParameter,
    repositoriesData,
    loading,
    called,
    error,
    repositoryNameSearchParameter,
  } = props;

  const hasNoQueryParameters: boolean = isNil(userLoginParameter) && isNil(repositoryNameSearchParameter);
  const isEmptyRepositoriesList: boolean = isEmpty(repositoriesData.elements);
  const totalAmount: number = useMemo((): number => {
    if (loading) {
      return -1;
    }

    if (hasNoQueryParameters || isEmptyRepositoriesList) {
      return 0;
    }

    return repositoriesData.repositoryCount;
  }, [hasNoQueryParameters, isEmptyRepositoriesList, loading, repositoriesData.repositoryCount]);

  return (
    <StyledPaper elevation={10}>
      <StyledTableContainer>
        <StyledTable stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {REPOSITORY_TABLE_COLUMNS.map(column => (
                <StyledTableCell key={column.id} align={column.align} style={{ minWidth: column.minWidth }}>
                  {column.label}
                </StyledTableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            <ContentRenderer
              isLoading={loading}
              isEmpty={hasNoQueryParameters || isEmptyRepositoriesList}
              emptyComponent={
                <EmptyTableBody
                  isEmptyRepositoriesList={isEmptyRepositoriesList}
                  hasNoQueryParameters={hasNoQueryParameters}
                  error={error}
                  called={called}
                />
              }
              loadingComponent={<LoadingTableBody />}
              contentComponent={
                <>
                  {repositoriesData.elements.map((repository: RepositoryResult) => (
                    <TableRow hover role="checkbox" tabIndex={-1} key={repository.id}>
                      {REPOSITORY_TABLE_COLUMNS.map(column => {
                        const value = `${repository[column.id]}`;
                        return (
                          <TableCell key={column.id} align={column.align}>
                            {repository.url && column.id === 'nameWithOwner' ? (
                              <Link href={repository.url} target="_blank" rel="noopener noreferrer">
                                {value}
                              </Link>
                            ) : (
                              value
                            )}
                          </TableCell>
                        );
                      })}
                    </TableRow>
                  ))}
                </>
              }
            />
          </TableBody>
        </StyledTable>
      </StyledTableContainer>
      <TablePagination
        rowsPerPageOptions={[50]}
        rowsPerPage={REPOSITORIES_PER_PAGE}
        component="div"
        count={totalAmount}
        page={page}
        onChangePage={handleChangePage}
      />
    </StyledPaper>
  );
};
