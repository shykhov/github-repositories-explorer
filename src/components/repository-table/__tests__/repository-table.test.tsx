import React from 'react';
import { render, shallow, ShallowWrapper } from 'enzyme';
import TablePagination from '@material-ui/core/TablePagination';
import 'jest-styled-components';

import { RepositoryTable, Props } from '../repository-table';

const repositoriesMockData = {
  repositoryCount: 3,
  pageInfo: {
    endCursor: 'Y3Vyc29yOjM=',
    hasNextPage: false,
    hasPreviousPage: false,
    startCursor: 'Y3Vyc29yOjE=',
  },
  elements: [
    {
      stars: 0,
      nameWithOwner: 'ASDQWEA1/ASDQWEA1',
      forks: 0,
      url: 'https://github.com/ASDQWEA1/ASDQWEA1',
      id: 'MDEwOlJlcG9zaXRvcnkxMTc2ODgyMTY=',
    },
    {
      stars: 0,
      nameWithOwner: 'JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd',
      forks: 0,
      url: 'https://github.com/JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd',
      id: 'MDEwOlJlcG9zaXRvcnkxNjQ1MTA5MDg=',
    },
    {
      stars: 0,
      nameWithOwner: 'SLCDevelopment/asdqweasdqweaaa',
      forks: 0,
      url: 'https://github.com/SLCDevelopment/asdqweasdqweaaa',
      id: 'MDEwOlJlcG9zaXRvcnkxNDg0NTcyNzQ=',
    },
  ],
};

describe('<RepositoryTable />', () => {
  let renderedComponent: unknown;
  let shallowRenderedComponent: ShallowWrapper;
  let propsMock: Props;

  beforeEach(() => {
    propsMock = {
      error: false,
      called: true,
      loading: false,
      repositoriesData: repositoriesMockData,
      handleChangePage: jest.fn(),
      page: 0,
    };

    renderedComponent = render(<RepositoryTable {...propsMock} />);
    shallowRenderedComponent = shallow(<RepositoryTable {...propsMock} />);
  });

  it('should render RepositoryTable', () => {
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should Pagination component have count property equal repositoryCount from props', () => {
    expect(shallowRenderedComponent.find(TablePagination).props().count).toEqual(
      propsMock.repositoriesData.repositoryCount,
    );
  });

  it('should Pagination component have count property equal -1 if is loading true', () => {
    shallowRenderedComponent.setProps({ loading: true });
    expect(shallowRenderedComponent.find(TablePagination).props().count).toEqual(-1);
  });
});
