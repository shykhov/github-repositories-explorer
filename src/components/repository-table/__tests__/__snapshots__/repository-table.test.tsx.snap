// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepositoryTable /> should render RepositoryTable 1`] = `
.c1 {
  max-height: 440px;
}

.c0 {
  max-width: 800px;
}

.c3 {
  font-weight: bold !important;
}

.c2 {
  table-layout: fixed;
}

<div
  class="MuiPaper-root c0 MuiPaper-elevation10 MuiPaper-rounded"
>
  <div
    class="MuiTableContainer-root c1"
  >
    <table
      aria-label="sticky table"
      class="MuiTable-root c2 MuiTable-stickyHeader"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head c3 MuiTableCell-stickyHeader"
            scope="col"
            style="min-width:170px"
          >
            Owner/Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head c3 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
            scope="col"
          >
            Stars
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head c3 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
            scope="col"
          >
            Forks
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabindex="-1"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="https://github.com/ASDQWEA1/ASDQWEA1"
              rel="noopener noreferrer"
              target="_blank"
            >
              ASDQWEA1/ASDQWEA1
            </a>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
        </tr>
        <tr
          class="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabindex="-1"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="https://github.com/JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd"
              rel="noopener noreferrer"
              target="_blank"
            >
              JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd
            </a>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
        </tr>
        <tr
          class="MuiTableRow-root MuiTableRow-hover"
          role="checkbox"
          tabindex="-1"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="https://github.com/SLCDevelopment/asdqweasdqweaaa"
              rel="noopener noreferrer"
              target="_blank"
            >
              SLCDevelopment/asdqweasdqweaaa
            </a>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="MuiTablePagination-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-3 of 3
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
